src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -269,6 +270,27 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
             return true;
         }
 
+        private const char HIGH_SURROGATE_START = '\ud800';
+        private const char HIGH_SURROGATE_END = '\udbff';
+        private const char LOW_SURROGATE_START = '\udc00';
+        private const char LOW_SURROGATE_END = '\udfff';
+
+        unsafe private static int PtrDiff(char* a, char* b)
+        {
+            return (int)(((uint)((byte*)a - (byte*)b)) >> 1);
+        }
+
+        // byte* flavor just for parity
+        unsafe private static int PtrDiff(byte* a, byte* b)
+        {
+            return (int)(a - b);
+        }
+
+        private static bool InRange(int ch, int start, int end)
+        {
+            return (uint)(ch - start) <= (uint)(end - start);
+        }
```

Comment:
It will make it harder to share the same implementation between different places though... .